File: Informatică și tehnologia informației/proba practica/rezolvare_2025/Subiectul_I/a.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <cmath>
#include <algorithm>

using namespace std;

// Un numar natural are un numar impar de divizori daca si numai daca este patrat perfect.
// Demonstratie: Divizorii unui numar se asociaza in perechi (d, x/d). 
// Singurul caz in care o pereche contine elemente egale este cand d = x/d, adica d^2 = x.
// Prin urmare, doar patratele perfecte au un divizor "nepereche", deci un numar total impar de divizori.
bool este_patrat_perfect(int x) {
    if (x < 0) return false;
    int r = round(sqrt(x));
    return r * r == x;
}

int main() {
    ifstream fin("divizori.in");
    ofstream fout("divizori.out");
    
    if (!fin) {
        cerr << "Eroare la deschiderea fisierului divizori.in" << endl;
        return 1;
    }
    
    int n;
    if (!(fin >> n)) return 0;
    
    vector<int> rez;
    for (int i = 0; i < n; ++i) {
        int x;
        fin >> x;
        if (este_patrat_perfect(x)) {
            rez.push_back(x);
        }
    }
    
    if (rez.empty()) {
        fout << "nu exista\n";
    } else {
        // Sortare descrescatoare
        sort(rez.begin(), rez.end(), greater<int>());
        for (size_t i = 0; i < rez.size(); ++i) {
            fout << rez[i] << (i == rez.size() - 1 ? "" : " ");
        }
        fout << "\n";
    }
    
    return 0;
}

File: Informatică și tehnologia informației/proba practica/rezolvare_2025/Subiectul_I/a.pas
program DivizoriA;
uses math;

var
  fin, fout: text;
  n, i, x, count: integer;
  rez: array[1..100] of integer;
  temp, j: integer;

// Un numar are un numar impar de divizori daca si numai daca este patrat perfect.
function EstePatratPerfect(x: integer): boolean;
var
  r: integer;
begin
  if x < 0 then
    EstePatratPerfect := false
  else
  begin
    r := round(sqrt(x));
    EstePatratPerfect := (r * r = x);
  end;
end;

begin
  assign(fin, 'divizori.in');
  reset(fin);
  assign(fout, 'divizori.out');
  rewrite(fout);
  
  if eof(fin) then
  begin
    close(fin);
    close(fout);
    exit;
  end;
  
  read(fin, n);
  count := 0;
  for i := 1 to n do
  begin
    read(fin, x);
    if EstePatratPerfect(x) then
    begin
      count := count + 1;
      rez[count] := x;
    end;
  end;
  
  if count = 0 then
    write(fout, 'nu exista')
  else
  begin
    // Sortare descrescatoare (Bubble Sort)
    for i := 1 to count - 1 do
      for j := i + 1 to count do
        if rez[i] < rez[j] then
        begin
          temp := rez[i];
          rez[i] := rez[j];
          rez[j] := temp;
        end;
        
    for i := 1 to count do
    begin
      write(fout, rez[i]);
      if i < count then
        write(fout, ' ');
    end;
    writeln(fout);
  end;
  
  close(fin);
  close(fout);
end.
